### PR TITLE
List pick component

### PR DESCRIPTION
**Pull request by ItsTisme**
## About The Pull Request

Adds a new component to let a user pick from a list. and pass the picked value as a result.

The player most be adjescent to the connected circuit for this to work.

## Why It's Good For The Game

You can do similar thing through voice activation, but this is honestly way easier for normal players to use instead of knowing "magic words" or having the circuit say all the options and copy pasting.

showcircuit

## Changelog
:cl:
add: New circuit component, list pick!
/:cl:

### CODE REVIEW

File: code/modules/wiremod/components/list/assoc_list_pick.dm [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+	circuit_flags = CIRCUIT_FLAG_INPUT_SIGNAL
+
+	var/index_type = PORT_TYPE_NUMBER
+
+/obj/item/circuit_component/assoc_list_pick/populate_options()
+	list_options = add_option_port("List Type", GLOB.wiremod_basic_types)
+
+/obj/item/circuit_component/assoc_list_pick/proc/make_list_port()
+	input_list = add_input_port("List", PORT_TYPE_ASSOC_LIST(PORT_TYPE_STRING,PORT_TYPE_LIST(PORT_TYPE_ANY)))
```
Comment (Mothblocks) [inline code/modules/wiremod/components/list/assoc_list_pick.dm:31]:
```suggestion
	input_list = add_input_port("List", PORT_TYPE_ASSOC_LIST(PORT_TYPE_STRING, PORT_TYPE_LIST(PORT_TYPE_ANY)))
```

File: code/modules/wiremod/components/list/assoc_list_pick.dm [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	if(port == list_options)
+		var/new_type = list_options.value
+		input_list.set_datatype(PORT_TYPE_ASSOC_LIST(PORT_TYPE_STRING,new_type))
+		output.set_datatype(new_type)
+
+
+/obj/item/circuit_component/assoc_list_pick/input_received(datum/port/input/port)
+	if(parent.Adjacent(user.value))
+		return
+	if(istype(user.value,/mob))
```
Comment (Mothblocks) [inline code/modules/wiremod/components/list/assoc_list_pick.dm:54]:
```suggestion

	if(ismob(user.value))
```

File: code/modules/wiremod/components/list/assoc_list_pick.dm [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+	var/datum/port/input/option/list_options
+
+	var/datum/port/input/input_list
+	var/datum/port/input/user
+	var/datum/port/input/input_name
+
+
+	var/datum/port/output/output
+	var/datum/port/output/succes
```
Comment (Mothblocks) [inline code/modules/wiremod/components/list/assoc_list_pick.dm:20]: `success`, two s

**Review by Mothblocks — CHANGES_REQUESTED** [review]
This seems really odd to me, I think it should be impossible to get this sort of prompt without any form of user input, which this would definitely make possible. I don't know how you would fix it cleanly but it's definitely a blocking concern. I think the best route for this is, instead of this opening up the list right away, it gives you a prompt (think like when being offered an item) that lets you click on it to then open up the input prompt. This prompt goes away when out of focus (right now you can pick a prompt after walking away from the circuit, you should not be able to).

**Comment by Autisem** [comment]
> This prompt goes away when out of focus (right now you can pick a prompt after walking away from the circuit, you should not be able to).

I'd have to rewrite partially how standard TGUI works or write my own wrapper for it, since the state is always on. I can do that


> instead of this opening up the list right away, it gives you a prompt (think like when being offered an item) that lets you click on it to then open up the input prompt.

I feel its kinda pointless, either way they will get a prompt. It also is not like you can just randomly open it you must be adjacent to the circuit containing it. if they don't want to pick they can just close it.

File: code/modules/wiremod/components/list/list_pick.dm [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		trigger_output.set_output(COMPONENT_SIGNAL)
+		INVOKE_ASYNC(src, PROC_REF(show_list), user.value, input_name.value, input_list.value)
+
+/obj/item/circuit_component/list_pick/proc/show_list(mob/user, message, list/showed_list)
+	if(!showed_list || showed_list.len == 0)
+		failure.set_output(COMPONENT_SIGNAL)
+		return
+	if(!message)
+		message = "circuit input"
+	var/picked = tgui_input_list(user, message = message, items = showed_list)
```
[the remark this answers is not in the document]
Comment (Autisem) [reply code/modules/wiremod/components/list/list_pick.dm:65]: I could have them both inherent from the same datum to have shared procs, but was not sure if that was anti-pattern for circuits or not

**Comment by Autisem** [comment]
Alternatively, I could either make it a chasy like keyboard or make it an upgrade you can attach to bot circuits or other stuff

File: code/modules/wiremod/components/list/list_pick.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/modules/wiremod/components/list/list_pick.dm:65]: Uhhh for something like this if it ends up being clean it should be ok

**Comment by Mothblocks** [comment]
> It also is not like you can just randomly open it you must be adjacent to the circuit containing it. 

Right but there's ways for circuits to be next to you without any input, especially for things like bots, it's crummy UX for us to be showing a prompt without any input

**Comment by Mothblocks** [comment]
Mmm they already have buttons and stuff so it might cause more problems than it solves but I think I could see it working if you're limited to one

**Comment by Autisem** [comment]
Hmm thinking about it I don't think it would work since it would make it too mutualy exclusive to other stuff(like what if I want to present a list in BCI?) not really sure how to adress the "UX" issue

**Comment by Autisem** [comment]
@Mothblocks I am now using a similair method that choice beacons use to check distance, the standard TGUI list pick proc does not account for state so unless I re-implent that entirly checking before and after would be the best way propably. I could also still implement the prompt of "do you want to see list input?" but I feel thats is more obnoxious then just showing the options list.

I could add a chat option of something bassicly saying "the circuit is showing a list of options"

File: code/modules/wiremod/components/list/assoc_list_pick.dm [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+/obj/item/circuit_component/assoc_list_pick/input_received(datum/port/input/port)
+	if(parent.Adjacent(user.value))
+		return
+
+	if(ismob(user.value))
+		trigger_output.set_output(COMPONENT_SIGNAL)
+		INVOKE_ASYNC(src, PROC_REF(show_list), user.value, input_name.value, input_list.value)
+
+/obj/item/circuit_component/assoc_list_pick/proc/show_list(mob/user, message, list/showed_list)
```
Comment (jlsnow301) [inline code/modules/wiremod/components/list/assoc_list_pick.dm:59]: please put a comment on this line

File: code/modules/wiremod/components/list/assoc_list_pick.dm [hunk 6]
```diff
@@ -0,0 +1,75 @@
+/**
+ * # associative List pick component
+ *
+ * Allows user to select 1 entry from a list
+ */
+
+/obj/item/circuit_component/assoc_list_pick
```
Comment (jlsnow301) [inline code/modules/wiremod/components/list/assoc_list_pick.dm:7]:
```suggestion
 */
/obj/item/circuit_component/assoc_list_pick
```

File: code/modules/wiremod/components/list/list_pick.dm [hunk 7]
```diff
@@ -0,0 +1,77 @@
+/**
+ * # List pick component
+ *
+ * Allows user to select 1 entry from a list
+ */
+
+/obj/item/circuit_component/list_pick
```
Comment (jlsnow301) [inline code/modules/wiremod/components/list/list_pick.dm:7]:
```suggestion
 */
/obj/item/circuit_component/list_pick
```

File: code/modules/wiremod/components/list/list_pick.dm [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+	var/datum/port/input/option/list_options
+
+	var/datum/port/input/input_list
+	var/datum/port/input/user
+	var/datum/port/input/input_name
+
+
+	var/datum/port/output/output
+	var/datum/port/output/succes
```
Comment (jlsnow301) [inline code/modules/wiremod/components/list/list_pick.dm:20]: please add comments above your vars for [Autodoc](https://github.com/tgstation/tgstation/blob/master/.github/guides/AUTODOC.md)
same applies for assoc list pick

File: code/modules/wiremod/components/list/assoc_list_pick.dm [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+	var/datum/port/input/option/list_options
+
+	var/datum/port/input/input_list
+	var/datum/port/input/user
+	var/datum/port/input/input_name
+
+
+	var/datum/port/output/output
+	var/datum/port/output/success
+	var/datum/port/output/failure
```
Comment (jlsnow301) [inline code/modules/wiremod/components/list/assoc_list_pick.dm:21]: [Autodoc](https://github.com/tgstation/tgstation/blob/master/.github/guides/AUTODOC.md) please

**Review by jlsnow301 — CHANGES_REQUESTED** [review]

File: code/modules/wiremod/components/list/assoc_list_pick.dm (hunk 9 above)
Comment (Autisem) [reply code/modules/wiremod/components/list/assoc_list_pick.dm]: Let me know if these comments are clear enough to other people

**Pull request merged**